Bug: Type inference fails for user-defined module function return types

## Description

Type inference does not work when calling functions from user-defined modules. The return type is not propagated to the variable, causing type mismatch errors when the variable is later used.

## Steps to Reproduce

**mylib.ez:**
```ez
module mylib

do get_value() -> u64 {
    return 42
}
```

**main.ez:**
```ez
module main

import & use @std
import & use lib"./mylib"

do needs_u64(v u64) {
    println("Got: ${v}")
}

do main() {
    temp val = lib.get_value()  // Type should be inferred as u64
    needs_u64(val)              // ERROR: expects u64, got <empty>
}
```

## Expected Behavior

`val` should be inferred as `u64` from `lib.get_value()`'s return type, and the code should compile successfully.

## Actual Behavior

```
error[E3001]: argument type mismatch in call to 'needs_u64': parameter 'v' expects u64, got 
  --> main.ez:13:15
   |
13 |     needs_u64(val)
   |               ^ types do not match
```

Note: "got " is empty — the type was never inferred.

## Root Cause

In `pkg/typechecker/typechecker.go`, the `inferModuleCallType()` function (line ~4704) handles stdlib modules but returns empty for user-defined modules:

```go
func (tc *TypeChecker) inferModuleCallType(member *ast.MemberExpression, args []ast.Expression) (string, bool) {
    switch moduleName {
    case "std":
        return tc.inferStdCallType(funcName, args)
    case "math":
        return tc.inferMathCallType(funcName, args)
    // ... other stdlib modules
    default:
        return "", false  // <-- User modules return empty type
    }
}
```

## Suggested Fix

The `default` case should use the existing `GetModuleFunction()` method to look up user module function signatures:

```go
default:
    // Check user-defined modules
    if sig, ok := tc.GetModuleFunction(moduleName, funcName); ok {
        if len(sig.ReturnTypes) >= 1 {
            return sig.ReturnTypes[0], true
        }
        return "void", true
    }
    return "", false
```

## Workaround

Explicitly declare the type:
```ez
temp val u64 = lib.get_value()  // Works with explicit type
```

## Additional Context

- Stdlib module calls work correctly: `temp ts = time.now()` infers `int`
- Same-module function calls work correctly
- Only user-defined module calls (via `import & use`) fail to infer types

--- a/integration-tests/pass/multi-file/user-module-type-inference/main.ez
+++ b/integration-tests/pass/multi-file/user-module-type-inference/main.ez
@@ -1,0 +1,48 @@
+/*
+ * main.ez - Test type inference for user-defined module function return types
+ *
+ * Tests issue #807: Type inference fails for user-defined module function return types
+ */
+module main
+
+import & use @std
+import & use lib"./mylib"
+
+do needs_u64(v u64) {
+    println("Got u64: ${v}")
+}
+
+do needs_string(s string) {
+    println("Got string: ${s}")
+}
+
+do needs_float(f float) {
+    println("Got float: ${f}")
+}
+
+do main() {
+    println("========================================")
+    println("  User Module Type Inference Test")
+    println("========================================")
+    println("")
+
+    // Test 1: Type inference from user module function returning u64
+    temp val = lib.get_value()  // Type should be inferred as u64
+    needs_u64(val)              // Should work without explicit type
+    println("Test 1 PASSED: u64 type inference")
+
+    // Test 2: Type inference from user module function returning string
+    temp str = lib.get_string()  // Type should be inferred as string
+    needs_string(str)
+    println("Test 2 PASSED: string type inference")
+
+    // Test 3: Type inference from user module function returning float
+    temp flt = lib.get_float()  // Type should be inferred as float
+    needs_float(flt)
+    println("Test 3 PASSED: float type inference")
+
+    println("")
+    println("========================================")
+    println("  All Tests PASSED!")
+    println("========================================")
+}

--- a/integration-tests/pass/multi-file/user-module-type-inference/mylib.ez
+++ b/integration-tests/pass/multi-file/user-module-type-inference/mylib.ez
@@ -1,0 +1,16 @@
+/*
+ * mylib.ez - Test module for type inference from user-defined modules
+ */
+module mylib
+
+do get_value() -> u64 {
+    return 42
+}
+
+do get_string() -> string {
+    return "hello"
+}
+
+do get_float() -> float {
+    return 3.14
+}

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -4725,6 +4725,13 @@ func (tc *TypeChecker) inferModuleCallType(member *ast.MemberExpression, args []
 	case "time":
 		return tc.inferTimeCallType(funcName, args)
 	default:
+		// Check user-defined modules
+		if sig, ok := tc.GetModuleFunction(moduleName, funcName); ok {
+			if len(sig.ReturnTypes) >= 1 {
+				return sig.ReturnTypes[0], true
+			}
+			return "void", true
+		}
 		return "", false
 	}
 }

